apps/webapp/app/runEngine/concerns/queues.server.ts

```diff
@@ -0,0 +1,245 @@
+import { sanitizeQueueName } from "@trigger.dev/core/v3/isomorphic";
+import { PrismaClientOrTransaction } from "@trigger.dev/database";
+import { AuthenticatedEnvironment } from "~/services/apiAuth.server";
+import { logger } from "~/services/logger.server";
+import { findCurrentWorkerFromEnvironment } from "~/v3/models/workerDeployment.server";
+import {
+  LockedBackgroundWorker,
+  QueueManager,
+  QueueProperties,
+  QueueValidationResult,
+  TriggerTaskRequest,
+} from "../types";
+import { WorkerGroupService } from "~/v3/services/worker/workerGroupService.server";
+import type { RunEngine } from "~/v3/runEngine.server";
+import { env } from "~/env.server";
+import { EngineServiceValidationError } from "./errors";
+
+export class DefaultQueueManager implements QueueManager {
+  constructor(
+    private readonly prisma: PrismaClientOrTransaction,
+    private readonly engine: RunEngine
+  ) {}
+
+  async resolveQueueProperties(
+    request: TriggerTaskRequest,
+    lockedBackgroundWorker?: LockedBackgroundWorker
+  ): Promise<QueueProperties> {
+    let queueName: string;
+    let lockedQueueId: string | undefined;
+
+    // Determine queue name based on lockToVersion and provided options
+    if (lockedBackgroundWorker) {
+      // Task is locked to a specific worker version
+      if (request.body.options?.queue?.name) {
+        const specifiedQueueName = request.body.options.queue.name;
+        // A specific queue name is provided
+        const specifiedQueue = await this.prisma.taskQueue.findFirst({
+          // Validate it exists for the locked worker
+          where: {
+            name: specifiedQueueName,
+            workers: { some: { id: lockedBackgroundWorker.id } }, // Ensure the queue is associated with any task of the locked worker
+          },
+        });
```

Comment:
@ericallam you should add the environmentId in here because the index on this table is name + runtimeEnvironmentId